style: apply /style-guide pass to sandboxes (#2668)
wandb/docs: sandboxes/lifecycle.mdx

---
title: Sandbox lifecycle
description: Learn about the lifecycle of a Serverless Sandbox, including its states, how to wait for state changes, and how to stop a sandbox.
keywords: ["SandboxState", "wait_until_complete", "wait_until_ready", "sandbox timeout", "resource cleanup"]
---

<Warning>
Serverless Sandboxes is in public preview.
</Warning>

This page describes the lifecycle of a Serverless Sandbox so that you can choose the right method for waiting on state changes, stopping a sandbox, and structuring code around its lifetime. Understanding the lifecycle helps you avoid race conditions, distinguish startup failures from runtime errors, and clean up resources predictably.

> A Serverless Sandbox goes through several states during its lifecycle. The sandbox's state determines which operations are available.

Reviewer: ```suggestion
A W&B Serverless Sandbox goes through several states during its lifecycle. The sandbox state determines which operations are available and how your code should wait for state changes, handle failures, stop the sandbox, and clean up resources.
```

Reviewer: Proposed felt a bit too wordy, imo.

Author: Applied a tightened version in 1159ffaa8: added W&B and dropped the possessive, but trimmed the longer clause since the paragraph just above already covers waiting, failures, and cleanup (per your too-wordy note).

| `CREATING`    | The container is provisioning. | 
| `RUNNING`     | The sandbox is ready for operations. | 
| `COMPLETED`   | The main process exited successfully with code `0`. | 
| `FAILED`      | The sandbox or its main process encountered an error during startup or execution. | 
| `TERMINATED`   | An external process stopped the sandbox, or it ended because it exceeded its maximum lifetime. |

Most operations require the sandbox to be in the `RUNNING` state. However, many operations wait until the sandbox is ready before proceeding.

After a sandbox enters a terminal state, it's no longer available for additional work.

## Wait for readiness or completion

The following sections describe the two methods for waiting on sandbox state changes and when to use each. Use different wait methods depending on what you need:

* Use `Sandbox.wait()` to [wait until the sandbox is ready for use](/sandboxes/lifecycle#wait-for-a-sandbox-to-start).
* Use `Sandbox.wait_until_complete()` to [wait until the sandbox's main process finishes](/sandboxes/lifecycle#wait-for-a-sandbox-to-complete).

### Wait for a sandbox to start

> Use `Sandbox.wait()` to wait for the sandbox to reach the `RUNNING` state. This method is useful for debugging startup problems or when you want to distinguish startup failures from errors in later commands.

Reviewer: ```suggestion
Use `Sandbox.wait()` to explicitly wait for the sandbox to reach the `RUNNING` state. This method is useful for debugging startup problems or when you want to distinguish startup failures from errors in later commands.
```

Author: Applied in 1159ffaa8.

## Explicit lifecycle control

> Use methods such as `Sandbox.wait()` and `Sandbox.stop()` when you need explicit control over the sandbox lifecycle instead of relying on automatic behavior.

Reviewer: ```suggestion
Use methods such as `Sandbox.wait()` and `Sandbox.stop()` when you need explicit control over the sandbox lifecycle.
```

Reviewer: I'd be careful of using "...automatic behavior." since it's contingent on loads of stuff, including whether or not the user set up their script right.

Author: Applied in 1159ffaa8; dropped the automatic-behavior phrasing per your note.
